hypervisor/arch/x86/guest/assign.c

```diff
@@ -187,14 +187,10 @@ ptirq_build_physical_rte(struct acrn_vm *vm, struct ptirq_remapping_info *entry)
 
 		/* init polarity & pin state */
 		if (rte.bits.intr_polarity == IOAPIC_RTE_INTPOL_ALO) {
-			if (entry->polarity == 0U) {
```

Comment:
every time when ptirq_alloc_entry, the entry will be memset to 0, so entry->polarity will be inited as 0. Not sure why there is redundant interrupt, could you help give more detail about the issue reproduce flow?

Reply:
The code here is not to fix the redundant  interrupt issue , it is to fix external interrupt lost issue.

I have a touchscreen device using IOAPIC direct IRQ interrupt with edge trigger mode (polarity is active low). 

1. During SOS boot,  ptirq_intx_pin_remap get called for the touchscreen  interrupt pin,  and this interrupt pin enry is allocated and  then maped  to SOS domain (no driver request this IRQ,  sos kernel IOAPIC driver behavior?), at this point , enrty->polarity is set to 1U. 

2. Then, i launch Anroid uos, and assign the touchscreen IRQ to uos using hypercall in acrn-dm, and pass the ACPI Interrupt information for the touchscrren device.

3.  During android uos boot ,  it write IOAPCI REDTBL  register(touchscreen driver requsest irq) , hv traps this action and vioapic_indirect_write get called, and here i notice vioapic_need_intr return true and an interrput  is inject into the uos (redundant  interrupt), 
and ptirq_build_physical_rte get called again ,  because entry->polariry is set to 1U, then vioapic_set_irqline_nolock is not called and the intial pin state is not set to the right state.

4.  Touchscreen controller trigger one interput(pull down the interrupt pin to low),  and in the hv interrupt flow ,  because the wrong pin state,  vioapic_set_pinstate will ignore the  first interrupt(old level is equal to new level).

Reply:
BTW,  even though entry->polarity is memset to zero when ptirq_alloc_entry, this initial value is meaningless, its value  shoud based on the rte.bits.intr_polarity, that is, after ` if (entry->polarity == 0U) {} and when entry->polarity = 1 or 0 ` (line 190~200 in assign.c),   the value become meaningful.

So we can't rely the value of entry->polarity  here.

Reply:
got it, it's because this ptirq entry first allocated to SOS then re-assign to UOS. The SOS set pin-state, and it become the init state for UOS - this is incorrect. How about just reset entry->polarity when we doing ptirq entry re-assignment? Like, in add_intx_remapping, add line 394?

390                 } else if (entry->vm != vm) {
391                         if (is_sos_vm(entry->vm)) {
392                                 entry->vm = vm;
393                                 entry->virt_sid.value = virt_sid.value;
394                                 entry->polarity = 0U;
395                         } else {

Reply:
Agree， new patch has been  pushed.